horovod/runner/gloo_run.py

```diff
@@ -280,7 +289,8 @@ def gloo_run(settings, nics, env, server_ip, command):
     # Each thread will use ssh command to launch the job on each remote host. If an
     # error occurs in one thread, entire process will be terminated. Otherwise,
     # threads will keep running and ssh session.
-    exec_command = _exec_command_fn(settings)
+    crash_event = threading.Event()
```

Comment:
Why are we passing this event into the `_exec_command_fn` method when we are not using it from the outside? We can easily create this event inside the method, which would keep the signature of this method unchanged and clean-up the changes inside the method a bit.

Enabling this behaviour could be controlled via settings. Should that even be the default / only useful behaviour?

@tgaddair I guess for Elastic Horovod, this needs to be turned off.

Reply:
> which would keep the signature of this method unchanged

Is this `_exec_command_fn` called more than the two times in this one file?  It seems to me like it is not.

> Enabling this behaviour could be controlled via settings

When I brought this up in slack, everybody thought this _was_ the only behavior.  The other behavior is just asking for hung workloads, I don't know why you would want that.

> I guess for Elastic Horovod, this needs to be turned off.

Isn't that already the case?  The codepath for `gloo_run_elastic`/`launch_gloo_elastic` should be unchanged.

> Why are we passing this event into the _exec_command_fn method when we are not using it from the outside?

I targeted the smallest change possible.  The only thing that seems broken is non-elastic gloo.  Is changing the `settings` something that other backends would have to accomodate?  I wasn't sure, so I avoided it.  But you're right, it could just as easily be a `def _execute_command_fn(settings, exit_after_worker_failure=False):`.

Reply:
> Isn't that already the case? The codepath for `gloo_run_elastic`/`launch_gloo_elastic` should be unchanged.

You are right, Elastic Horovod uses its own `_exec_command_fn`.

Then this is the only useful behaviour, great!

Can you move that event into the `_exec_command_fn`? This will clean up things there as well (e.g. the `if crash_event:`).